precompiles/utils/src/revert.rs

```diff
@@ -0,0 +1,352 @@
+			match (i, part) {
+				(0, BacktracePart::Field(field)) => write!(f, "{field}")?,
+				(_, BacktracePart::Field(field)) => write!(f, ".{field}")?,
+				(_, BacktracePart::Tuple(index)) => write!(f, ".{index}")?,
+				(_, BacktracePart::Array(index)) => write!(f, "[{index}]")?,
+			}
+		}
+		Ok(())
+	}
+}
+
+/// Possible revert reasons.
+#[non_exhaustive]
+#[derive(PartialEq, Eq)]
+pub enum RevertReason {
+	/// A custom revert reason if other variants are not appropriate.
+	Custom(String),
+	/// Tried to read data out of bounds.
+	ReadOutOfBounds {
+		/// What was being read?
+		what: String,
+	},
+	/// An unknown selector has been provided.
+	UnknownSelector,
+	/// A value is too large to fit in the wanted type.
+	/// For security reasons integers are always parsed as `uint256` then
+	/// casted to the wanted type. If the value overflows this type then this
+	/// revert is used.
+	ValueIsTooLarge {
+		/// What was being read?
+		what: String,
+	},
+	/// A pointer (used for structs and arrays) points out of bounds.
+	PointerToOutofBound,
+	/// The reading cursor overflowed.
+	/// This should realistically never happen as it would require an input
+	/// of length larger than 2^64, which would cost too much to be included
+	/// in a block.
+	CursorOverflow,
+	/// Used by a check that the input contains at least N static arguments.
+	/// Often use to return early if the input is too short.
+	ExpectedAtLeastNArguments(usize),
+}
+
+impl RevertReason {
+	/// Create a `RevertReason::Custom` from anything that can be converted to a `String`.
+	/// Argument is the custom revert message.
+	pub fn custom(s: impl Into<String>) -> Self {
+		RevertReason::Custom(s.into())
+	}
+
+	/// Create a `RevertReason::ReadOutOfBounds` from anything that can be converted to a `String`.
+	/// Argument names what was expected to be read.
+	pub fn read_out_of_bounds(what: impl Into<String>) -> Self {
+		RevertReason::ReadOutOfBounds { what: what.into() }
+	}
+
+	/// Create a `RevertReason::ValueIsTooLarge` from anything that can be converted to a `String`.
+	/// Argument names what was expected to be read.
+	pub fn value_is_too_large(what: impl Into<String>) -> Self {
+		RevertReason::ValueIsTooLarge { what: what.into() }
+	}
+}
+
+impl core::fmt::Display for RevertReason {
+	fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> Result<(), core::fmt::Error> {
+		match self {
+			RevertReason::Custom(s) => write!(f, "{s}"),
+			RevertReason::ReadOutOfBounds { what } => {
+				write!(f, "Tried to read {what} out of bounds")
+			}
+			RevertReason::UnknownSelector => write!(f, "Unknown selector"),
+			RevertReason::ValueIsTooLarge { what } => write!(f, "Value is too large for {what}"),
+			RevertReason::PointerToOutofBound => write!(f, "Pointer points to out of bound"),
+			RevertReason::CursorOverflow => write!(f, "Reading cursor overflowed"),
+			RevertReason::ExpectedAtLeastNArguments(n) => {
+				write!(f, "Expected at least {n} arguments")
```

Comment:
this write is not taking any argument, is it correct?

Reply:
`write!` macro can use variable names in templates directly.
`write!("{n}")` is equivalent to `write!("{}", n)`

Reply:
Pretty cool, never saw this before in Rust

Reply:
It's pretty "recent", wasn't in the Book when I learned Rust but now it is :p